crates/network-scanner-net/src/runtime.rs

```diff
@@ -81,130 +79,99 @@ impl Socket2Runtime {
     pub(crate) fn remove_socket(&self, socket: &socket2::Socket, id: usize) -> anyhow::Result<()> {
         self.poller.delete(socket)?;
         // remove all events related to this socket
-        self.event_cache.lock().retain(|event| id == event.0.key);
+        self.remove_events_with_id_from_history(id);
         Ok(())
     }
 
     fn start_loop(
         &self,
         register_receiver: Receiver<RegisterEvent>,
-        event_sender: Sender<Event>,
+        event_history: Arc<DashSet<EventWrapper>>,
     ) -> anyhow::Result<()> {
-        // To prevent an Arc cycle with the Arc<Socket2Runtime>, we added additional indirection.
-        // Otherwise, the reference in the new thread would prevent the runtime from being dropped and shutdown.
-        let poller = self.poller.clone();
+        // we make is_terminated Arc<AtomicBool> and poller Arc<Poller> so that we can clone them and move them into the thread
+        // we cannot hold a Arc<Socket2Runtime> in the thread, because it will create a cycle reference and the runtime will never be dropped.
```

Comment:
style: I encourage writing proper sentences starting with an uppercase and ending with a period.